### ISSUE TITLE
Cannot blit from single-plane image to single plane of multi-plane image

### ISSUE THREAD

**Issue by haasn** [KhronosGroup/Vulkan-Docs#2029]
As per the current spec,

[ VUID-VkImageBlit2-aspectMask-00238](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkImageBlit2-aspectMask-00238):
>  The `aspectMask` member of `srcSubresource` and `dstSubresource` **must** match.

This prevents a blit from e.g. an R8 image to a single plane of an G8_R8_B8_3PLANE_444 image, an operation that seems entirely reasonable to perform (and, indeed, can be emulated trivially using a raster or compute shader).

It seem that the rules around this VUID could probably be relaxed around planar images.

**Comment by spencer-lunarg** [comment]
You are also going to run into ` VUID-vkCmdBlitImage-dstImage-06422`

> dstImage must not use a [format that requires a sampler Y′CBCR conversion](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#formats-requiring-sampler-ycbcr-conversion)

**Comment by zlatinski** [comment]
We have purposefully disallowed the use of the Blit and clear operations since they call for complex color space conversion with filtering and may result in unpredictable (unexpected) results.

Multi-planar format cannot be set as a render target for similar reasons (with the exception of a few add-on extensions allowing for that). To accomplish that, it would be necessary to set up an output sampler that would specify how a color in the RGB color space would be transformed to YCbCr. Although it is not impossible, I doubt there has been sufficient demand for this to become an extension.

**Comment by haasn** [comment]
To be clear, I imagine that blitting `vec4(0.2, 0.4, 0.6, 0.8)` to plane 1 of an `G8_R8B8_2PLANE` would fill it with `vec2(0.2, 0.4)` values, **not* anything YCbCr-converted.

**Comment by fluppeteer** [comment]
When Y'CbCr was specified, the expectation was that you would create a G8_R8B8_2PLANE image with VK_IMAGE_CREATE_DISJOINT_BIT set, then create images of single-plane formats over the separate planes with VK_IMAGE_CREATE_ALIAS_BIT. Then you should be able to blit to and from the individual planes - the expectation that this was the approach is why we didn't expand the blit functionality at the time. However, that does depend on the implementation exposing VK_FORMAT_FEATURE_DISJOINT_BIT for the relevant format. It may still be useful to expose ways to blit to/from individual aspects, but I wanted to make sure the alias approach was known in case it solves the problem in the short term.

**Comment by HansKristian-Work** [comment]
I think this is a good opportunity to consider adding Blit/Clear with either support for VkImageView or explicit format provided in copy_commands2. I agree this is a spec hole.

Main awkward part of blit/clear on planar images right now is how the formats work. Blits and clear work on the image format itself, not any view format, and plane 0 is technically G, so it would get confusing how component mapping would work. Is clear color G or R used? On a blit for plane 1, is there a swizzle to/from B8R8? To make it more usable, we would need explicit formats to make it obvious how that should work I think.

Another workaround would be taking plane views with VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT. If STORAGE is supported, it would work to implement dstBlit through storage image. srcBlit could work with sampled image.

Worst case if nothing is supported is to blit to/from a separate image and use vkCmdCopyImage to move in/out of the YCbCr image, but that's quite inefficient.

**Comment by fluppeteer** [comment]
> Main awkward part of blit/clear on planar images right now is how the formats work. Blits and clear work on the image format itself, not any view format, and plane 0 is technically G, so it would get confusing how component mapping would work. Is clear color G or R used?

Yes - that's one reason we side-stepped the issue by allowing aliases (in this case of an R8 image over the G8 plane). If you actually wanted a vkBlitWithAspects I imagine you'd want to go green-to-green - but it does raise the problem that blitting blue-and-red to the chroma channels is incompatible with the luma/G plane if they're downsampled.

**Comment by HansKristian-Work** [comment]
> if they're downsampled

If VK_IMAGE_ASPECT_n_BIT is used there is a unique extent. Same as for VkImageView of planes, the image extent changes based on that. VK_IMAGE_ASPECT_COLOR_BIT blits and clears on YCbCr is bogus I think.

**Comment by HansKristian-Work** [comment]
> vkCmdClearColorImage / vkCmdClearAttachments were intended to work on planar imageviews as well

vkCmdClearAttachments work on image views, so this is surprising. Is there a specific VU that bans this? If anything, I suspect it talks about COLOR_ASPECT being the only thing allowed with PLANE_ASPECT having been missed.

**Comment by fluppeteer** [comment]
Agreed. It just kind of messes with the logic that you might want to map R, G and B aspects from a single (4:4:4) colour image - which just says that's the wrong way to think about it. I think I'm persuaded that people may not have be creating their own YCbCr images, and therefore requiring that they be created with certain flags may be a problem. And I guess blitting to depth/stencil is related. So I have no problem with vkCmdBlitView or something, but it does mean that some code will need to support more than one code path.

**Comment by fluppeteer** [comment]
> vkCmdClearAttachments work on image views, so this is surprising. Is there a specific VU that bans this? If anything, I suspect it talks about COLOR_ASPECT being the only thing allowed with PLANE_ASPECT having been missed.

Cross-over. I don't remember, and I think most of the YCbCr stuff predated VUs, so it could absolutely have fallen through the cracks.

**Comment by pdaniell-nv** [comment]
I've created an internal issue [here](https://gitlab.khronos.org/vulkan/vulkan/-/issues/3335) to investigate the new image view blit/clear extension.

**Comment by HansKristian-Work** [comment]
> it doesn't need to be expanded

Blitting in Vulkan is a fairly complex in the sense that formats are expanded, filtering can be applied and format conversion happens on store. It's more or less a builtin way of doing sampling and a render pass (or compute pass depending on what the implementation feels like) with builtin image views.

> copy the values

vkCmdCopyImage would cover that case already I think. It really sounds like what you want here is raw copy into the planes? Is the scaling + filtering + format load/store part of the copy (which blit provides) required here?

If we want to just add blit/clear to planes without any new API entry points or pNexts, it might possible to inherit wording from vkCmdCopyImage.
Based on vkCmdCopyImage wording around planar images, it talks about https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#formats-compatible-planes when copying between planes. It might be possible to just reuse that wording for blits and clears.

**Comment by haasn** [comment]
> > vkCmdClearColorImage / vkCmdClearAttachments were intended to work on planar imageviews as well
> 
> vkCmdClearAttachments work on image views, so this is surprising. Is there a specific VU that bans this? If anything, I suspect it talks about COLOR_ASPECT being the only thing allowed with PLANE_ASPECT having been missed.

* For each element of pAttachments, the corresponding attachment in the current render pass instance must either not be backed by an image view, or contain each of the aspects specified in aspectMask

You have to specify `COLOR_ASPECT` to clear a color attachment, which then fails this VU because the corresponding attachment is backed by an image not containing the color aspect.

**Comment by HansKristian-Work** [comment]
> [VUID-vkCmdClearAttachments-pAttachments-07270](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCmdClearAttachments-pAttachments-07270): For each element of pAttachments, the corresponding attachment in the current render pass instance must either not be backed by an image view, or contain each of the aspects specified in aspectMask

The intention here is that it's the view that needs to contain the aspect. Pretty sure it was an oversight and can be fixed.

**Comment by spencer-lunarg** [comment]
issue update: The VUID-vkCmdClearAttachments-pAttachments-07270 VU was updated in the [1.3.242 spec update](https://github.com/KhronosGroup/Vulkan-Docs/commit/f3e686fc22251102713a2a1a1cd071f60ab6efd8#diff-51b3a7d73dfaaf16f2b4cfc74503c6e8da60f8a391edba14cd6f3172bb0a014e)
